src/_pytest/_code/source.py

```diff
@@ -305,11 +310,7 @@ def getfslineno(obj) -> Tuple[Optional[Union["Literal['']", py.path.local]], int
                 _, lineno = findsource(obj)
             except IOError:
                 pass
```

Comment:
@bluetech 
We've discussed the `fspath = fn and py.path.local(fn) or None` line above recently.
I think it might actually never happen (and will try to write a test for it), but in general I think we should remove the `Optional` from the return type, which would simpify typing internally already.

Reply:
This is relevant for https://github.com/pytest-dev/pytest/blob/0e777a6ef5ef28ac830116a16fbc047af202c86a/src/_pytest/compat.py#L310-L317

Reply:
It depends on whether `fn = inspect.getsourcefile(obj) or inspect.getfile(obj)` can end up being falsey. I didn't check.